filament/src/Material.cpp

```diff
@@ -64,20 +68,28 @@ BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder const
 BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder&& rhs) noexcept = default;
 
 Material::Builder& Material::Builder::package(const void* payload, size_t size) {
-    mImpl->mPayload = payload;
-    mImpl->mSize = size;
+    mImpl->mBufferDescriptor.reset(new BufferDescriptor(payload, size));
+    return *this;
+}
+
+Material::Builder& Material::Builder::package(BufferDescriptor&& buffer) {
+    *mImpl->mBufferDescriptor = std::move(buffer);
     return *this;
 }
 
 Material* Material::Builder::build(Engine& engine) {
-    MaterialParser* materialParser = new MaterialParser(
-            upcast(engine).getBackend(), mImpl->mPayload, mImpl->mSize);
+    // By moving the stashed buffer descriptor into a temporary, we are ensuring that its
+    // user-supplied callback is called when execution leaves this scope.
+    BufferDescriptor bd = std::move(*mImpl->mBufferDescriptor);
```

Comment:
I think this doesn't work if:

- you did a copy of the builder. The shared-pointer will track the buffer correctly, but this line will destroy the buffer right here.
- you called `.build()` more than once